src/appservice/AppService.ts

```diff
@@ -0,0 +1,150 @@
+            onUserQuery: appService.onUserQuery.bind(appService),
+            onEvent: appService.onEvent.bind(appService),
+        };
+        return appService;
+    }
+
+    /**
+     * Start the appservice for the end user with the appropriate settings from their config and registration file.
+     * @param port The port to make the appservice listen for transactions from the homeserver on (usually sourced from the cli).
+     * @param config The parsed configuration file.
+     * @param registrationFilePath A path to their homeserver registration file.
+     */
+     public static async run(port: number, config: IConfig, registrationFilePath: string) {
+        const dataStore = new PgDataStore(config.db.connectionString);
+        await dataStore.init();
+        const service = await MjolnirAppService.makeMjolnirAppService(config, dataStore, registrationFilePath);
+        // Can't stress how important it is that listen happens last.
+        await service.start(port);
+    }
+
+    public onUserQuery (queriedUser: MatrixUser) {
+        return {}; // auto-provision users with no additonal data
+    }
+
+    /**
+     * Handle an individual event pushed by the homeserver to us.
+     * This function is async (and anything downstream would be anyway), which does mean that events can be processed out of order.
+     * Not a huge problem for us, but is something to be aware of.
+     * @param request A matrix-appservice-bridge request encapsulating a Matrix event.
+     * @param context Additional context for the Matrix event.
+     */
+    public async onEvent(request: Request<WeakEvent>, context: BridgeContext) {
+        const mxEvent = request.getData();
+        if ('m.room.member' === mxEvent.type) {
+            if ('invite' === mxEvent.content['membership'] && mxEvent.state_key === this.bridge.botUserId) {
+                await this.mjolnirManager.provisionNewMjolnir(mxEvent.sender);
+            }
+        }
+        this.accessControl.handleEvent(mxEvent['room_id'], mxEvent);
+        this.mjolnirManager.onEvent(request, context);
+    }
+
+    /**
+     * Start the appservice. See `run`.
+     * @param port The port that the appservice should listen on to receive transactions from the homeserver.
+     */
+    private async start(port: number) {
+        console.log("Starting MjolnirAppService, Matrix-side to listen on port %s", port);
+        this.api.start(this.config.webAPI.port);
+        await this.bridge.listen(port);
+        console.log("MjolnirAppService started successfully");
+    }
+
+    /**
+     * Stop listening to requests from both the homeserver and web api and disconnect from the datastore.
+     */
+    public async close(): Promise<void> {
+        await this.bridge.close();
+        await this.dataStore.close();
+        await this.api.close();
+    }
+
+    public static generateRegistration(reg: AppServiceRegistration, callback: (finalRegisration: AppServiceRegistration) => void) {
```

Comment:
Not very clear what this method does. Do I understand correctly that it's to create a registration file for initial setup?

Also, typo `finalRegis**t**ration`.